include/onnxruntime/core/framework/data_types.h

```diff
@@ -353,6 +355,92 @@ constexpr ONNX_NAMESPACE::TensorProto_DataType ToTensorDataType<BFloat16>() {
   return ONNX_NAMESPACE::TensorProto_DataType_BFLOAT16;
 }
 
+template <typename T>
```

Comment:
I hate to be painful, but is all this worth it just to get a typename in a few errors when the build has RTTI disabled? This sort of error should be rare, and if it occurred the user can be asked to re-run the model with a full build. 

For the Cast error the int value for the type (stored in Cast::to_) could be output to provide good enough information.